### PR TITLE
update findStatCommand, fix redis

### PR DESCRIPTION
**Pull request by ndoriomedov**
http://outsourcing.nat.tepkom.ru:20080/redmine/issues/1011

### CODE REVIEW

File: src/TimeService.Broker/Requests/CompanyService.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        companiesData =
+          (await RequestHandler.ProcessRequest<IGetCompaniesRequest, IGetCompaniesResponse>(
+            _rcGetCompanies,
+            IGetCompaniesRequest.CreateObj(usersIds),
+            errors,
+            _logger))
+          ?.Companies;
+      }
+
+      if (companiesData is null)
```
Comment (MariyaSin) [inline src/TimeService.Broker/Requests/CompanyService.cs:54]: user no need know about our problems

File: src/TimeService.Broker/Requests/DepartmentService.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+          IGetDepartmentUsersRequest.CreateObj(
+            departmentId,
+            skipCount: skipCount,
+            takeCount: takeCount),
+          errors,
+          _logger);
+
+      if (response is null)
+      {
+        errors.Add("Can not get department users data.");
```
Comment (MariyaSin) [inline src/TimeService.Broker/Requests/DepartmentService.cs:50]: same

File: src/TimeService.Broker/Requests/DepartmentService.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        departmentsData =
+          (await RequestHandler.ProcessRequest<IFilterDepartmentsRequest, IFilterDepartmentsResponse>(
+            _rcFilterDepartments,
+            IFilterDepartmentsRequest.CreateObj(departmentsIds),
+            errors,
+            _logger))
+          ?.Departments;
+      }
+
+      if (departmentsData is null)
```
Comment (MariyaSin) [inline src/TimeService.Broker/Requests/DepartmentService.cs:78]: same

File: src/TimeService.Broker/Requests/ProjectService.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        projectsData =
+          (await RequestHandler.ProcessRequest<IGetProjectsRequest, IGetProjectsResponse>(
+            _rcGetProjects,
+            IGetProjectsRequest.CreateObj(projectsIds, includeUsers: true),
+            errors,
+            _logger))
+          ?.Projects;
+      }
+
+      if (projectsData is null)
```
Comment (MariyaSin) [inline src/TimeService.Broker/Requests/ProjectService.cs:59]: same

File: src/TimeService.Broker/Requests/ProjectService.cs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+          (await RequestHandler.ProcessRequest<IGetProjectsUsersRequest, IGetProjectsUsersResponse>(
+            _rcGetProjectsUsers,
+            IGetProjectsUsersRequest.CreateObj(
+              projectsIds: new() { projectId },
+              skipCount: skipCount,
+              takeCount: takeCount),
+            errors,
+            _logger));
+
+      if (response is null)
```
Comment (MariyaSin) [inline src/TimeService.Broker/Requests/ProjectService.cs:79]: same

File: src/TimeService.Broker/Requests/UserService.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        usersData =
+          (await RequestHandler.ProcessRequest<IGetUsersDataRequest, IGetUsersDataResponse>(
+            _rcGetUsersData,
+            IGetUsersDataRequest.CreateObj(usersIds),
+            errors,
+            _logger))
+          ?.UsersData;
+      }
+
+      if (usersData is null)
```
Comment (MariyaSin) [inline src/TimeService.Broker/Requests/UserService.cs:54]: same

File: src/TimeService.Broker/Requests/DepartmentService.cs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+namespace LT.DigitalOffice.TimeService.Broker.Requests
+{
+  public class DepartmentService : IDepartmentService
+  {
+    private readonly ILogger<DepartmentService> _logger;
+    private readonly IGlobalCacheRepository _globalCache;
+    private readonly IRequestClient<IFilterDepartmentsRequest> _rcFilterDepartments;
+    private readonly IRequestClient<IGetDepartmentUsersRequest> _rcGetDepartmentUsers;
+    private readonly IRequestClient<IGetDepartmentsRequest> _rcGetDepartments;
```
Comment (MariyaSin) [inline src/TimeService.Broker/Requests/DepartmentService.cs:24]: not assigned

File: src/TimeService.Broker/Requests/Interfaces/IDepartmentService.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+{
+  [AutoInject]
+  public interface IDepartmentService
+  {
+    Task<(List<Guid> usersIds, int totalCount)> GetDepartmentUsersAsync(
+      Guid departmentId,
+      List<string> errors,
+      int? skipCount = null,
+      int? takeCount = null,
+      DateTime? byEntryDate = null);
```
Comment (MariyaSin) [inline src/TimeService.Broker/Requests/Interfaces/IDepartmentService.cs:17]: add empty line

File: src/TimeService.Broker/Requests/Interfaces/IDepartmentService.cs [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    Task<List<DepartmentData>> GetDepartmentsDataAsync(
+      List<string> errors,
+      List<Guid> departmentsIds = null,
+      List<Guid> usersIds = null,
+      List<Guid> projectsIds = null);
```
Comment (MariyaSin) [inline src/TimeService.Broker/Requests/Interfaces/IDepartmentService.cs:22]: add empty line

File: src/TimeService.Broker/Requests/Interfaces/IProjectService.cs [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+{
+  [AutoInject]
+  public interface IProjectService
+  {
+    Task<List<ProjectData>> GetProjectsDataAsync(
+      List<string> errors,
+      List<Guid> projectsIds = null,
+      Guid? departmentId = null,
+      Guid? userId = null,
+      bool includeUsers = false);
```
Comment (MariyaSin) [inline src/TimeService.Broker/Requests/Interfaces/IProjectService.cs:17]: add empty line

File: src/TimeService.Broker/Requests/UserService.cs [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+      int takecount,
+      List<string> errors,
+      bool? ascendingSort = null)
+    {
+      if (usersIds is null || !usersIds.Any())
+      {
+        return default;
+      }
+
+      return default; // fix in next commit
```
Comment (MariyaSin) [inline src/TimeService.Broker/Requests/UserService.cs:69]: you can use FindUsersConsumer from userService

File: src/TimeService.Business/Commands/Import/ImportStatCommand.cs [hunk 12]
```diff
@@ -48,227 +36,75 @@ public class ImportStatCommand : IImportStatCommand
...
-        _logger.LogWarning(logError + "Errors: {errors}.", string.Join("\n", result.Message.Errors));
-      }
-      catch (Exception exc)
-      {
-        _logger.LogError(exc, logError);
-      }
-
-      errors.Add(messageError);
-      return null;
+      return await _projectService.GetProjectsDataAsync(errors, projectsIds: projectIds, includeUsers: includeUsers);
```
Comment (MariyaSin) [inline src/TimeService.Business/Commands/Import/ImportStatCommand.cs:60]: you can get projectswithout private method

File: src/TimeService.Business/Commands/Import/ImportStatCommand.cs [hunk 13]
```diff
@@ -48,227 +36,75 @@ public class ImportStatCommand : IImportStatCommand
...
     }
 
     private async Task<List<ProjectUserData>> GetProjectsUsersAsync(List<Guid> usersIds, List<string> errors)
     {
-      string messageError = "Cannot get projects users info. Please, try again later.";
-      const string logError = "Cannot get projects users info.";
-
-      if (usersIds == null || !usersIds.Any())
+      if (usersIds is null || !usersIds.Any())
```
Comment (MariyaSin) [inline src/TimeService.Business/Commands/Import/ImportStatCommand.cs:65]: same

File: src/TimeService.Business/Commands/Import/ImportStatCommand.cs [hunk 14]
```diff
@@ -48,227 +36,75 @@ public class ImportStatCommand : IImportStatCommand
...
-      }
-      catch (Exception exc)
-      {
-        _logger.LogError(exc, logError, departmentId);
-      }
-
-      errors.Add(messageError);
-
-      return null;
+      return (await _departmentService.GetDepartmentUsersAsync(departmentId, errors, byEntryDate: filterbyEntryDate)).usersIds;
```
Comment (MariyaSin) [inline src/TimeService.Business/Commands/Import/ImportStatCommand.cs:75]: same

File: src/TimeService.Business/Commands/Import/ImportStatCommand.cs [hunk 15]
```diff
@@ -48,227 +36,75 @@ public class ImportStatCommand : IImportStatCommand
...
-        return usersFromCache;
-      }
-
-      return await GetUsersDataFromBrokerAsync(usersIds, errors);
-    }
-
-    private async Task<List<UserData>> GetUsersDataFromBrokerAsync(List<Guid> usersIds, List<string> errors)
-    {
-      if (usersIds == null || !usersIds.Any())
+      if (usersIds is null || !usersIds.Any())
```
Comment (MariyaSin) [inline src/TimeService.Business/Commands/Import/ImportStatCommand.cs:80]: same

File: src/TimeService.Business/Commands/Import/ImportStatCommand.cs [hunk 16]
```diff
@@ -48,227 +36,75 @@ public class ImportStatCommand : IImportStatCommand
...
-      errors.Add(message);
-
-      return null;
+      return await _userService.GetUsersDataAsync(usersIds, errors);
     }
 
     private async Task<List<DepartmentData>> GetDapartmentsAsync(List<Guid> projectsIds, List<string> errors)
     {
-      if (projectsIds == null || !projectsIds.Any())
+      if (projectsIds is null || !projectsIds.Any())
```
Comment (MariyaSin) [inline src/TimeService.Business/Commands/Import/ImportStatCommand.cs:90]: same

File: src/TimeService.Business/Commands/Import/ImportStatCommand.cs [hunk 17]
```diff
@@ -48,227 +36,75 @@ public class ImportStatCommand : IImportStatCommand
...
+      return await _departmentService.GetDepartmentsDataAsync(errors, projectsIds: projectsIds);
     }
 
-    private async Task<List<DepartmentData>> GetDepartmentsFromBrokerAsync(List<Guid> projectsIds, List<string> errors)
+    private async Task<List<CompanyData>> GetCompaniesAsync(
+      List<Guid> usersIds,
+      List<string> errors)
     {
-      if (projectsIds == null || !projectsIds.Any())
+      if (usersIds is null || !usersIds.Any())
```
Comment (MariyaSin) [inline src/TimeService.Business/Commands/Import/ImportStatCommand.cs:102]: same

File: src/TimeService.Business/Commands/LeaveTime/FindLeaveTimesCommand.cs [hunk 18]
```diff
@@ -43,126 +34,32 @@ public class FindLeaveTimesCommand : IFindLeaveTimesCommand
...
+    private readonly IUserService _userService;
+    private readonly ICompanyService _companyService;
     private readonly IResponseCreator _responsCreator;
 
     #region private methods
 
     private async Task<List<UserData>> GetUsersData(List<Guid> usersIds, List<string> errors)
     {
-      if (usersIds == null || !usersIds.Any())
+      if (usersIds is null || !usersIds.Any())
```
Comment (MariyaSin) [inline src/TimeService.Business/Commands/LeaveTime/FindLeaveTimesCommand.cs:45]: same

File: src/TimeService.Business/Commands/LeaveTime/FindLeaveTimesCommand.cs [hunk 19]
```diff
@@ -43,126 +34,32 @@ public class FindLeaveTimesCommand : IFindLeaveTimesCommand
...
-      return null;
+      return await _userService.GetUsersDataAsync(usersIds, errors);
     }
 
     private async Task<List<CompanyData>> GetCompaniesAsync(
       List<Guid> usersIds,
       List<string> errors)
     {
-      if (usersIds == null || !usersIds.Any())
+      if (usersIds is null || !usersIds.Any())
```
Comment (MariyaSin) [inline src/TimeService.Business/Commands/LeaveTime/FindLeaveTimesCommand.cs:57]: same

File: src/TimeService.Business/Commands/WorkTime/FindWorkTimesCommand.cs [hunk 20]
```diff
@@ -43,202 +35,48 @@ public class FindWorkTimesCommand : IFindWorkTimesCommand
...
-      return await GetProjectsThroughBroker(projectsIds, userId, errors);
-    }
-
-    private async Task<List<ProjectData>> GetProjectsThroughBroker(List<Guid> projectsIds, Guid? userId, List<string> errors)
-    {
-      string messageError = "Cannot get projects info. Please, try again later.";
-      const string logError = "Cannot get projects info.";
-
-      if (projectsIds == null || !projectsIds.Any())
+      if (projectsIds is null || !projectsIds.Any())
```
Comment (MariyaSin) [inline src/TimeService.Business/Commands/WorkTime/FindWorkTimesCommand.cs:52]: same

File: src/TimeService.Business/Commands/WorkTime/FindWorkTimesCommand.cs [hunk 21]
```diff
@@ -43,202 +35,48 @@ public class FindWorkTimesCommand : IFindWorkTimesCommand
...
-        return usersFromCache;
-      }
-
-      return await GetUsersDataFromBrokerAsync(usersIds, errors);
-    }
-
-    private async Task<List<UserData>> GetUsersDataFromBrokerAsync(List<Guid> usersIds, List<string> errors)
-    {
-      if (usersIds == null || !usersIds.Any())
+      if (usersIds is null || !usersIds.Any())
```
Comment (MariyaSin) [inline src/TimeService.Business/Commands/WorkTime/FindWorkTimesCommand.cs:62]: same

**Review by VLiamin — APPROVED** [review]

**Pull request merged**